### PR TITLE
Updated version constraint copy

### LINKED ISSUE #4160: Clarify caret syntax
### What information is missing?

As per the [docs of caret syntax for dependencies](https://dart.dev/tools/pub/dependencies#caret-syntax), it states that:
* `^1.2.3` is equivalent to `'>=1.2.3 <2.0.0'` (up to the next **major** release, exclusive)
* `^0.1.2` is equivalent to `'>=0.1.2 <0.2.0'` (up to the next **minor** release, exclusive)

What is the correct interpretation here ?
Does it mean that `^0.1.2` extends to the next **minor** because:
* if and only if it is `0.x.x` (based on current **major** being `< 1`)
* if and only if it is `x.1.x` (based on current **minor** being `< 2`)

Also, as the question might arise later on:
* what would `^0.0.0` translates into ?
* what would `^2.0.0` translates into ?

This issue relates and complements #168 / #193 / #825 / #1075

### How would you like us to fix this problem?

I think you should just mention which criteria will make the caret syntax extends to the next **major** release (exclusive) or the next **minor** release (exclusive).

### PR DESCRIPTION
**Pull request by atsansone**
Rewrote details on version constraints. Fixes #4160

Staged: https://dart--staging--dart-mrtpv5g3.web.app/tools/pub/dependencies#version-constraints

### CODE REVIEW

**Comment by MaryaBelanger** [comment]
@atsansone can you stage this?

File: src/tools/pub/dependencies.md [hunk 1]
```diff
@@ -460,9 +462,9 @@ this flag only resolves new dependencies if:
 * `pubspec.lock` is not missing
 * The packages' [content hashes][] match
 
-[`--enforce-lockfile`]: /tools/pub/cmd/pub-get#--enforce-lockfile
-[lockfile]: /tools/pub/glossary#lockfile
-[content hashes]: /tools/pub/glossary#content-hashes
+[`--enforce-lockfile`]: {{site.url}}/tools/pub/cmd/pub-get#--enforce-lockfile
```
Comment (parlough) [inline src/tools/pub/dependencies.md:465]: I haven't taken a look at this PR yet, but just scanning quickly, we don't use `{{site.url}}` on dart.dev. Just the base relative URLs.

**Comment by parlough** [comment]
Consider getting someone on the pub team to review this as well once it is staged.

File: src/tools/pub/dependencies.md [hunk 2]
```diff
@@ -16,13 +16,14 @@ At the end is a list of
...
 ## Overview
 
-For each dependency, you specify the *name* of the package you depend on
-and the *range of versions* of that package that you allow.
-You can also specify the
-[*source*](/tools/pub/glossary#source),
-which tells pub how to locate the package.
+For each dependency, you specify the _name_ of the package you depend on
+and the _range of versions_ of that package that you allow.
+You can also specify the [_source_][].
```
Comment (jonasfj) [inline src/tools/pub/dependencies.md:21]:
```suggestion
You can also specify the [_source_].
```

is this not sufficient?

File: src/tools/pub/dependencies.md [hunk 3]
```diff
@@ -16,13 +16,14 @@ At the end is a list of
...
+The source tells pub how to locate the package.
 
-Here's an example of specifying a dependency:
+[_source_]: {{site.url}}/tools/pub/glossary#source
+
+To specify a dependency:
```
Comment (jonasfj) [inline src/tools/pub/dependencies.md:26]: The old text does make clear that this is an example.

To me it wrong to claim that this how an example

File: src/tools/pub/dependencies.md [hunk 4]
```diff
@@ -237,93 +235,101 @@ If it's an unknown identifier, the dependency is always considered unsatisfied.
...
-The Dart community uses semantic versioning<sup id="fnref:semver"><a
-href="#fn:semver">1</a></sup>, which helps you know which versions should work.
-If you know that your package works fine with `1.2.3` of some dependency, then
-semantic versioning tells you that it should work with any subsequent
-stable release before `2.0.0`.
-For details on pub's version system,
-see the [package versioning page](/tools/pub/versioning#semantic-versions).
+Let's say that your Package A depends upon Package B.
+How can you communicate to other developers which version of Package B
+has been verified with a given version of Package A?
```
Comment (jonasfj) [inline src/tools/pub/dependencies.md:240]: verified is a strong claim here.

File: src/tools/pub/dependencies.md [hunk 5]
```diff
@@ -237,93 +235,101 @@ If it's an unknown identifier, the dependency is always considered unsatisfied.
...
+| `<=1.2.3` | Given version or earlier                |    No    |                                                                                                                                                         |
+|  `<1.2.3` | Versions earlier than the given version |    No    | Use this when you know an upper bound version that _doesn't_ work with your package. This version might be the first to introduce some breaking change. |
+{:.table}
 
-`1.2.3`
-: A concrete version number pins the dependency to only allow that
-    <em>exact</em> version. Avoid using this when you can because it can cause
-    version lock for your users and make it hard for them to use your package
-    along with other packages that also depend on it.
+You can specify any combination of version values as their ranges intersect.
```
Comment (szakarias) [inline src/tools/pub/dependencies.md:289]: What it is exactly we mean here?

File: src/tools/pub/dependencies.md (hunk 4 above)
Comment (szakarias) [reply src/tools/pub/dependencies.md:240]: Maybe use 'compatible'?

**Review by szakarias — APPROVED** [review]
LGTM with the comments

File: src/tools/pub/dependencies.md (hunk 2 above)
Comment (atsansone) [reply src/tools/pub/dependencies.md:21]: It is [sufficient](https://kramdown.gettalong.org/syntax.html#reference-links), but would be inconsistent with other links in the corpus.

File: src/tools/pub/dependencies.md (hunk 5 above)
Comment (atsansone) [reply src/tools/pub/dependencies.md:289]: I have a better explanation in the updated copy.

File: src/tools/pub/dependencies.md (hunk 1 above)
Comment (atsansone) [reply src/tools/pub/dependencies.md]: True, and I will remove. I think we should going forward to be consistent between both sites.

**Review by atsansone — COMMENTED** [review]
@jonasfj , @szakarias , @parlough : Made updates per your comments.

File: src/tools/pub/dependencies.md (hunk 1 above)
Comment (parlough) [reply src/tools/pub/dependencies.md]: That's the goal. I hope to make it no longer necessary on flutter/website soon.

**Review by jonasfj — APPROVED** [review]

**Pull request merged**